### ISSUE TITLE
Does DeepJDOT works for regression problem?

### ISSUE THREAD

**Issue by offchan42** [bbdamodaran/deepJDOT#5]
@bbdamodaran I want to do regression from 0 to 1. (Not 0 or 1), how do I modify the code to do so? Suppose I already have the data available in the deepjdot_demo by replacing target values with 0.9 or 0.1 instead of 1 or 0. One of the changes I can do to the demo is to predict how close the sample is to the center.
My real use case is to predict how much someone opens their mouth or how much their open their eyes (0 = completely closed, 1 = completely opened) with source domain as image from day lighting and target domain as image from night lighting (and maybe from different person face).

If deepjdot doesn't work with regression, what are some of the other approaches I can use?

**Comment by bbdamodaran** [comment]
It should be possible with deepjdot. I will look into this may be on next week

**Comment by offchan42** [comment]
If I want to check this myself (because it's urgent for me), can I just feed the target values with 0 to 1 instead of 0 or 1? (And then change softmax to sigmoid for multiple regression values)
Does deepjdot do something specific for classification (like loss function that only works for classification)?

Please guide me where to investigate/change (don't need to tell exactly what to change). Is it about loss function or model structure? Something like that.

I will explain how I think about changing it. Normally classification with multi-class will use `categorical_crossentropy` loss. I would want to change it to `binary_crossentropy` so that it works with multi-output regression.
But I see that you write your own loss function. If I just want to change the bit about `categorical` to `binary`, how do I do it?

**Comment by offchan42** [comment]
I saw that you used 2 loss functions.
![image](https://user-images.githubusercontent.com/15215732/60027138-14045f80-96c7-11e9-8d00-763e820568bc.png)
I am sure I need to modify `classifier_cat_loss`. Do I also need to modify `align_loss` for intermediate layers?

**Comment by bbdamodaran** [comment]
DeepJDOT will work on classification and regression tasks. DeepJDOT uses two loss functions: Classification loss and feature alignment loss. For the regression you need to just change the loss function (regression loss function), and also set class_balance as false.

The alignment loss is general, and no need to change for regression tasks

**Comment by bbdamodaran** [comment]
did you succeed with your the problem?

**Comment by offchan42** [comment]
I've been learning about your demo code. I was able to reproduce the result from SVHN and MNIST with accuracy of MNIST around 84% (I stopped training the code early around a few hundred iterations instead of training for 15,000 iterations). I think that if I keep training I will reach 90% like you did. It's impressive what your algorithm can do there! The accuracy without jdot was only 65% on the MNIST.

But even the 65% result before domain adaptation is still impressing me. I'm quite mesmerized and suspicious about it because the random guesser would be able to do only 10% accuracy. But the model has learned to predict MNIST even though it hasn't seen MNIST even once? That's good to know the model has somewhat generalization ability without domain adaptation. I thought it was going to perform so bad like a random guess.

What I want to do with regression first is to do a small twist to the SVHN to MNIST demo to make sure that something bad does not happen. I want to change the task from digit classification to angle regression. I want to rotate each image from -90 to 90 degrees and let model predict 0 to 1 corresponding to the angle rotated and see if JDOT is able to give less MeanAbsoluteError. What do you think about this test?

Am I correct that I need to just change this line to `binary_crossentropy`?
![image](https://user-images.githubusercontent.com/15215732/60198433-c4ab6400-986b-11e9-9f53-84b48bc1a69a.png)
And this line to False?
![image](https://user-images.githubusercontent.com/15215732/60198478-e1479c00-986b-11e9-8b25-d18cd65a344f.png)


I also have a few questions about hyper-parameter tuning for JDOT. Is there a general practice to do it or I don't need to change it even if I change the model structure? Like those `sloss`, `tloss`, `jdot_alpha` stuff.

**Comment by offchan42** [comment]
Hi @bbdamodaran , I tried creating the model that predicts how much the digit is rotated from -90 to 90 degrees. (0 to 1 regression) but I'm stuck with the problem that the target dataset's MeanAbsoluteError is not improved when training with DeepJDOT.

Could you try tinker with my fork and make the MAE(mean-absolute-error) on target data improve? I duplicated the file `deep_jdot_svhn_mnist.py` to be `deep_jdot_svhn_mnist_2.py` and change it to do regression. So you just need to run that 2nd file using Spyder. Run cell by cell to see what I changed. You can try doing `diff` check between `deep_jdot_svhn_mnist.py` and `deep_jdot_svhn_mnist_2.py` to see what I changed also.
https://github.com/off99555/deepJDOT

I don't quite understand your algorithm so it is very difficult for me to debug this issue myself. So if you can get your hands on it I would appreciate it very much. That way you will know practically that your algorithm works with regression also.

PS. I swapped the source data to be MNIST and target data to be SVHN at the `generate_rotated_images` function call. You can change it back if you want. But I think it's better this way because the random rotation causes SVHN to provide black pixels, but MNIST is already black everywhere so it will be more difficult for the model to overfit.
PS2. I changed the model structure to be a little smaller to experiment faster.
PS3. I changed image normalization from gaussian to 0to1 min max (so that I can plot the image).

Please let me know if you have the time to check it.

**Comment by offchan42** [comment]
I think that you are busy so I was trying to solve the issue on my own. You don't need to check it anymore.
I found that the problem was mainly 2 folds:
1. I did not train the source model long enough. It seems that the source model's weight is quite important as the initializer for the target model. (Not sure though)
2. The dataset differs too much. So I changed the max random rotation angle from 90 to 45 and it worked. (Very sure about this effect)

I now have low mean absolute error on the target testset. Before jdot, the target testset MAE was around 0.138, after jdot it was around 0.024 after 750 iterations. The source testset MAE was around 0.014.

# Conclusion
DeepJDOT works for regression with some limitations on how different the source and target can be. Next, I'll try it on my own dataset and report the result.

PS. I removed the `deepjdot_svhn_mnist_2.py` and put both classification and regression functionality into `deepjdot_svhn_mnist.py` instead.
